docbrown/db/src/graph_window.rs

```diff
@@ -48,367 +50,393 @@ pub struct WindowedGraph {
+    where
+        S: Send + 'static,
+        F: Fn(VertexRef) -> S + Send + Sync + Copy,
+        F2: Fn(S, S) -> S + Sync + Send + Copy,
+    {
+        self.graph
+            .fold_window_par(self.actual_start(t_start), self.actual_end(t_end), f, agg)
     }
 
-    fn in_degree(self) -> Self::ValueIterType<usize> {
-        Box::new(self.map(|v| v.in_degree()))
+    fn edge_ref<V1: Into<VertexRef>, V2: Into<VertexRef>>(
+        &self,
+        src: V1,
+        dst: V2,
+    ) -> Option<EdgeRef> {
+        self.graph
+            .edge_ref_window(src, dst, self.t_start, self.t_end)
     }
 
-    fn out_degree(self) -> Self::ValueIterType<usize> {
-        Box::new(self.map(|v| v.out_degree()))
+    fn edge_ref_window<V1: Into<VertexRef>, V2: Into<VertexRef>>(
+        &self,
+        src: V1,
+        dst: V2,
+        t_start: i64,
+        t_end: i64,
+    ) -> Option<EdgeRef> {
+        self.graph
+            .edge_ref_window(src, dst, self.actual_start(t_start), self.actual_end(t_end))
     }
 
-    fn edges(self) -> Self::EList {
-        Box::new(self.flat_map(|v| v.edges()))
+    fn edge_refs(&self) -> Box<dyn Iterator<Item = EdgeRef> + Send> {
+        self.graph.edge_refs_window(self.t_start, self.t_end)
     }
 
-    fn in_edges(self) -> Self::EList {
-        Box::new(self.flat_map(|v| v.in_edges()))
+    fn edge_refs_window(
+        &self,
+        t_start: i64,
+        t_end: i64,
+    ) -> Box<dyn Iterator<Item = EdgeRef> + Send> {
+        self.graph
+            .edge_refs_window(self.actual_start(t_start), self.actual_end(t_end))
     }
 
-    fn out_edges(self) -> Self::EList {
-        Box::new(self.flat_map(|v| v.out_edges()))
+    fn vertex_edges(&self, v: VertexRef, d: Direction) -> Box<dyn Iterator<Item = EdgeRef> + Send> {
+        self.graph
+            .vertex_edges_window(v, self.t_start, self.t_end, d)
     }
 
-    fn neighbours(self) -> Self {
-        Box::new(self.flat_map(|v| v.neighbours()))
+    fn vertex_edges_window(
+        &self,
+        v: VertexRef,
+        t_start: i64,
+        t_end: i64,
+        d: Direction,
+    ) -> Box<dyn Iterator<Item = EdgeRef> + Send> {
+        self.graph
+            .vertex_edges_window(v, self.actual_start(t_start), self.actual_end(t_end), d)
     }
 
-    fn in_neighbours(self) -> Self {
-        Box::new(self.flat_map(|v| v.in_neighbours()))
+    fn vertex_edges_window_t(
+        &self,
+        v: VertexRef,
+        t_start: i64,
+        t_end: i64,
+        d: Direction,
+    ) -> Box<dyn Iterator<Item = EdgeRef> + Send> {
+        self.graph
+            .vertex_edges_window_t(v, self.actual_start(t_start), self.actual_end(t_end), d)
     }
 
-    fn out_neighbours(self) -> Self {
-        Box::new(self.flat_map(|v| v.out_neighbours()))
+    fn neighbours(&self, v: VertexRef, d: Direction) -> Box<dyn Iterator<Item = VertexRef> + Send> {
+        self.graph.neighbours_window(v, self.t_start, self.t_end, d)
```

Comment:
Once we fix the window bounds, do these two (neighbours and neighbours_window) end up basically with the same semantics? I imagine we need to keep both to make the trait happy?

Reply:
Note that neighbours_window allows you to restrict the view further, may be useful for chaining views...